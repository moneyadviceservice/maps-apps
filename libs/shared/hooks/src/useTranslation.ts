import { ReactNode } from 'react';
import { TranslationGroup } from './types';
import useLanguage from './useLanguage';

/*
The useTranslation hook that handles all translation inside this app.

Why?

We needed something quick and easy that could handle two languages (only ever two),
and also showed the English right there in the file.

This makes it very easy to find and change text.
It also follows along with the whole tailwind approach of having everything inside the
same file, which is nice.

We did look at other libraries, but not too many of them store the text inside the component.
Also, these libraries are designed for many languages, where we only ever have two.

So, we chose this approach for now. It might seem like a terrible idea, but it should
be easy to replace, and is very easy to understand.

Update Sep 2024: added t() method to load translations from locale json files.

Applications are free to choose between the two methods. Mix them if you like.

Example uses:

(1) Stick to `z()` for shared components where you want to the text to be consistent
for every consumer of the component.

(2) Use the `t()` method when string might be amended by content editors (eg as a sudo CMS).

Usage:

```
// Inline text
const { z } = useTranslation()
return ( <h1> {z({en: "Hello World", cy: "Hello World"})} </h1> )

// Object interpolation
const { z } = useTranslation()
return ( <h1> {z({en: "Hello {thing}", cy: "Helo {thing}"}, { thing: "world"})} </h1> )

// JSX
const { z } = useTranslation()
return ( <h1> {z({en: <div>hello</div>, cy: <div>helo</div>})} </h1> )

// Locale json file
const { t } = useTranslation()
return ( <h1> {t("helloWorld")} </h1> )

// Locale json with interpolation
const { t } = useTranslation()
return ( <h1> {t("helloWorld", { thing: "world"})} </h1> )

```

If there is no cy translation then just leave it out, and some console errors will be shown.
You can then hand these over for translation.
*/

type Data = Record<string, string>;
const fallbackText = '<No translation available>';

type FlattenJsonOptions = {
  data: Record<string, unknown>;
  parentKey?: string;
  result?: Record<string, string>;
};

const flattenJson = ({
  data,
  parentKey = '',
  result = {},
}: FlattenJsonOptions): Record<string, string> => {
  for (const key in data) {
    if (Object.prototype.hasOwnProperty.call(data, key)) {
      const newKey = parentKey ? `${parentKey}.${key}` : key;

      // If the value is an object, recursively flatten it
      const isObject =
        data[key] && typeof data[key] === 'object' && !Array.isArray(data[key]);
      if (isObject) {
        flattenJson({
          data: data[key] as Record<string, unknown>,
          parentKey: newKey,
          result,
        });
      } else {
        result[newKey] = String(data[key]);
      }
    }
  }
  return result;
};

export const useTranslation = (localeOverirde?: string) => {
  let locale = useLanguage();
  if (localeOverirde) {
    locale = localeOverirde;
  }

  if (!(locale === 'en' || locale === 'cy')) {
    throw new Error(`locale [${locale}] does not exist`);
  }

  const interpolate = ((node: ReactNode, data?: Data) => {
    if (typeof node !== 'string') {
      return node;
    }
    if (!data) {
      return node;
    }
    return Object.keys(data).reduce((previousValue, currentValue) => {
      if (typeof previousValue !== 'string') return node;
      return previousValue.replaceAll(`{${currentValue}}`, data[currentValue]);
    }, node);
  }) as <T extends ReactNode>(
    node: T,
    data?: Data,
  ) => T extends string ? string : T;

  const findFirstAvailableTranslation = (translation: TranslationGroup) => {
    const firstKey = Object.keys(translation)[0];
    return translation[firstKey as keyof TranslationGroup];
  };

  const z = ((
    translation: TranslationGroup,
    data?: Data,
    fallback = fallbackText,
  ) => {
    const value = translation[locale as keyof TranslationGroup];
    if (value) {
      return interpolate(value, data);
    }

    console.warn(
      `no translation available in [${locale}] for [${findFirstAvailableTranslation(
        translation,
      )}]`,
    );
    return fallback;
  }) as <T extends TranslationGroup>(
    translation: T,
    data?: Data,
    fallback?: string,
  ) => T[keyof T] extends string ? string : T[keyof T];

  const t = (key: string, data?: Data, fallback = fallbackText) => {
    try {
      // import the translations from the applications /public/locales folder
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const translations = require(`/public/locales/${locale}.json`);
      const value = flattenJson({ data: translations })[key];

      if (value) {
        return interpolate(value, data);
      }

      return fallback;
    } catch (e) {
      console.warn(
        `no translation available in /public/locales/${locale}.json for [${key}]`,
      );
      return fallback;
    }
  };

  return {
    z,
    t,
    locale,
  } as const;
};

export default useTranslation;
