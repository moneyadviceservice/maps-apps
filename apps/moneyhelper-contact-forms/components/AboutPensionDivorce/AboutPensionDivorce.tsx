import useTranslation from '@maps-react/hooks/useTranslation';

import { StepComponent } from '../../lib/types';
import { FormWrapper } from '../FormWrapper';
import { SectionsRenderer } from '../SectionsRenderer';
import { SubTitleRenderer } from '../SubTitleRenderer';

export const AboutPensionDivorce: StepComponent = ({ step }) => {
  const { t, tList } = useTranslation();
  const sections = tList('components.about-pension-divorce.sections');

  return (
    <div className="flex flex-col gap-4">
      <SubTitleRenderer
        content={t('components.about-pension-divorce.sub-title')}
        testId="about-pension-divorce-sub-title"
      />
      <SectionsRenderer
        sections={sections}
        testIdPrefix="about-pension-divorce"
      />
      <FormWrapper step={step}></FormWrapper>
    </div>
  );
};
