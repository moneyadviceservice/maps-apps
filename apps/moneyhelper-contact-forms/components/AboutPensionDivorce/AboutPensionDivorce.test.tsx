import { render } from '@testing-library/react';

import useTranslation from '@maps-react/hooks/useTranslation';

import { AboutPensionDivorce } from './AboutPensionDivorce';

jest.mock('@maps-react/hooks/useTranslation');

const mockUseTranslation = useTranslation as jest.Mock;

// Mock the `useTranslation` hook
describe('AboutPensionDivorce Component', () => {
  beforeEach(() => {
    mockUseTranslation.mockReturnValue({
      t: (key: string) => key,
      tList: () => {
        return [
          {
            title: 'mock-title',
            content: 'mock-content',
            items: ['mock-item', 'mock-item'],
            footer: 'mock-footer',
          },
        ];
      },
    });
  });

  it('renders component correctly', () => {
    const { container } = render(<AboutPensionDivorce step="mock-step" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
