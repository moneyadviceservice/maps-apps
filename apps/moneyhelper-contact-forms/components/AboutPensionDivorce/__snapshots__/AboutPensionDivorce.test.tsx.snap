// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AboutPensionDivorce Component renders component correctly 1`] = `
<div
  class="flex flex-col gap-4"
>
  <p
    class="text-xl md:text-2xl text-gray-800 pt-4 mb-2 font-light"
    data-testid="about-pension-divorce-sub-title"
  >
    components.about-pension-divorce.sub-title
  </p>
  <section
    class="flex flex-col gap-4"
  >
    <h2
      class="text-2xl md:text-4xl text-gray-800 font-bold"
      data-testid="components-about-pension-divorce-title-0"
    >
      mock-title
    </h2>
    <p
      class="mb-2"
      data-testid="paragraph"
    >
      mock-content
    </p>
    <div
      class="flex items-center"
      data-testid="list-element"
    >
      <ul
        class="marker:mr-2 marker:pr-2 space-y-2 marker:leading-snug list-disc marker:text-blue-700 mb-3 ml-8"
        data-testid="about-pension-divorce-list-0"
        start="-1"
      >
        <li>
          <p
            class="mb-4"
            data-testid="paragraph"
          >
            mock-item
          </p>
        </li>
        <li>
          <p
            class="mb-4"
            data-testid="paragraph"
          >
            mock-item
          </p>
        </li>
      </ul>
    </div>
    <p
      class="mb-4"
      data-testid="components-about-pension-divorce-footer-0"
    >
      mock-footer
    </p>
  </section>
  <form
    action="/api/form-handler"
    aria-label="Contact form for Step: mock-step"
    method="POST"
    novalidate=""
  >
    <input
      name="lang"
      type="hidden"
      value=""
    />
    <button
      class="border-0 text-white bg-magenta-500 hover:bg-pink-800 active:bg-pink-900 active:outline-yellow-400 active:text-white t-button relative px-3 py-2 font-semibold rounded outline-none shadow-bottom-gray inline-flex items-center justify-between gap-2 cursor-pointer disabled:cursor-not-allowed disabled:border-slate-400 disabled:border disabled:bg-transparent disabled:text-gray-400 disabled:shadow-none focus:bg-yellow-400 focus:outline-blue-700 focus:outline-[3px] focus:outline-offset-0 focus:text-gray-800 focus:shadow-none focus:before:content-[""] focus:before:w-[calc(100%+6px)] focus:before:rounded-[0.45rem] focus:before:absolute focus:before:inset-0 focus:before:inset-x-[-3px] focus:before:inset-y-[-3px] focus:before:shadow-bottom-gray active:outline-[3px] active:outline-offset-0 active:shadow-inset active:before:content-[""] active:before:w-[calc(100%+6px)] active:before:rounded-[0.45rem] active:before:absolute active:before:inset-0 active:before:inset-x-[-3px] active:before:inset-y-[-3px] active:before:shadow-bottom-white t-primary-button w-full mt-10 mb-4 md:w-auto"
      data-testid="continue-button"
      type="submit"
    >
      <span />
      common.form.button.continue-text
      <span />
    </button>
  </form>
</div>
`;
